RuminantFarmSystems/RuFaS
# Time refactor

Restructured the Time class data:
1. Got rid of the two-dimensional array: years: List[List[int]]
    1. The `years` variable is turned into a property for now (along with `start_year_int`, `start_day`, `end_year_int`, `end_day`), as the Weather class is still heavily dependent on this structure. These properties will be removed following the refactoring of the Weather class as they are no longer needed.
2. Use a single DateTime object to keep track of the current date instead.
3. Created the following properties for ease of use:
    1. current_julian_day: the current Julian day of the current year in integer.
    2. current_month: the current month in integer.
    3. current_simulation_year: the current simulation year in integer.
    4. current_calendar_year: the current calendar year in integer.
    5. simulation_day: the current simulation day in integer.
    6. year_start_day: the first Julian day of the current year in integer.
    7. year_end_day: the last Julian day of the current year in integer.
4. Updated all other functions in the Time class accordingly.
5. Updated all references to the Time class.

Other changes:
1. Moved `_record_animal_events()`, `_record_heiferIIs_conception_rate()`, and `_record_cows_conception_rate()` from `AnimalManager` into `AnimalModuleReporter` as they serve the purpose of reporting the herd data at the end of the simulation.
### Context

Issue(s) closed by this pull request: closes #1522 & #1533.

      
How big are the changes in the PR? Would you nominate it for a major version upgrade?

### What

See description.

### Why

See description.

### How

See description.

### Test plan

1. Update all affected tests.
2. Run the simulation to check if all the days within the simulation range are reached. This also fixes the bug where there was a day missing for the previous Time class.
